File: back/src/main/java/com/backend/luaspets/Services/CartService.java
package com.backend.luaspets.Services;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.backend.luaspets.Model.Cart;
import com.backend.luaspets.Model.CartItem;
import com.backend.luaspets.Repository.CartItemRepository;
import com.backend.luaspets.Repository.CartRepository;

@Service
public class CartService {
    
     @Autowired
    private CartRepository cartRepository;

    @Autowired
    private CartItemRepository cartItemRepository;

    // Método para crear un nuevo carrito
    public Cart createCart(Cart cart) {
        return cartRepository.save(cart);
    }

    // Método para obtener un carrito por su ID
    public Cart getCartById(Integer cartId) {
        return cartRepository.findById(cartId).orElse(null);
    }

    // Método para agregar un producto al carrito
    public CartItem addProductToCart(Integer cartId, CartItem cartItem) {
        // Verificar si el carrito existe
        Cart cart = cartRepository.findById(cartId)
                .orElseThrow(() -> new RuntimeException("Cart not found"));

        // Establecer la relación entre el carrito y el producto
        cartItem.setCart(cart);
        return cartItemRepository.save(cartItem);
    }

    // Método para obtener todos los productos de un carrito específico
    public List<CartItem> getAllItemsInCart(Integer cartId) {
        return cartItemRepository.findAll().stream()
                .filter(item -> item.getCart().getIdCart().equals(cartId))
                .toList();
    }

    // Método para actualizar la cantidad de un producto en el carrito
    public CartItem updateProductQuantity(Integer cartItemId, Integer quantity) {
        CartItem cartItem = cartItemRepository.findById(cartItemId)
                .orElseThrow(() -> new RuntimeException("CartItem not found"));

        cartItem.setQuantity(quantity);
        return cartItemRepository.save(cartItem);
    }

    // Método para eliminar un producto del carrito
    public void removeProductFromCart(Integer cartItemId) {
        cartItemRepository.deleteById(cartItemId);
    }

}

File: front/src/app/services/cart/cart.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { catchError, Observable } from 'rxjs';
import { Cart } from 'src/app/models/cart.model';
import { CartItem } from 'src/app/models/cartItem.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CartService {

  private baseUrl = 'http://localhost:8080/api/cart'; // URL base de la API

  constructor(private http: HttpClient) {}

  // Método para crear un nuevo carrito
  createCart(cart: Cart): Observable<Cart> {
    return this.http.post<Cart>(`${this.baseUrl}`, cart);
  }

  // Método para obtener un carrito por su ID
  getCartById(cartId: number): Observable<Cart> {
    return this.http.get<Cart>(`${this.baseUrl}/${cartId}`);
  }

  // Método para agregar un producto al carrito
  addItemToCart(url: string) {
    return this.http.post(url, null); // Envía la solicitud POST con la URL generada
  }

  // Método para obtener todos los productos de un carrito
  getAllItemsInCart(cartId: number): Observable<CartItem[]> {
    return this.http.get<CartItem[]>(`${this.baseUrl}/${cartId}/items`);
  }

  // Método para actualizar la cantidad de un producto en el carrito
  updateProductQuantity(cartItemId: number, quantity: number): Observable<CartItem> {
    const body = { quantity };
    return this.http.put<CartItem>(`${this.baseUrl}/items/${cartItemId}`, body);
  }

  // Método para eliminar un producto del carrito
  removeProductFromCart(cartItemId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/items/${cartItemId}`);
  }
}
